Possible bug in checking @EnsureCalledMethods

Hi,

I have this code (relevant part of the class):

```java
public class ResultStore implements Closeable {

    private final @Owning ResourcesCollection<ChronicleQueue> queues;
    private final @Owning ResourcesCollection<ExcerptAppender> appenders;
    private final @Owning ChronicleQueue queryStoreQueue;
    private final @Owning ExcerptAppender queryStoreAppender;

    public ResultStore(List<File> resultPaths, File queryFilePath)
    {
        queues = new ResourcesCollection<>(() -> new ArrayList<>(resultPaths.size()));
        resultPaths.forEach(path -> queues.add(SingleChronicleQueueBuilder.single(path.toJavaIOFile()).build()));
        appenders = new ResourcesCollection<>(() -> new ArrayList<>(resultPaths.size()));
        queues.forEach(q -> appenders.add(q.acquireAppender()));
        queryStoreQueue = queryFilePath != null ? SingleChronicleQueueBuilder.single(queryFilePath.toJavaIOFile()).build() : null;
        queryStoreAppender = queryStoreQueue != null ? queryStoreQueue.acquireAppender() : null;
    }

    @EnsuresCalledMethods.List({
        @EnsuresCalledMethods(value = { "queryStoreAppender", "queryStoreQueue" }, methods = { "close" }),
        @EnsuresCalledMethods(value = { "appenders", "queues" }, methods = { "removeAll" })
    })
    @Override
    public void close()
    {
        appenders.removeAll(Closeable::close);
        if (queryStoreAppender != null)
            queryStoreAppender.close();
        queues.removeAll(Closeable::close);
        if (queryStoreQueue != null)
            queryStoreQueue.close();
    }
}
```

which fails with:
```
    [javac] Compiling 1 source file to /home/jlewandowski/dev/cassandra/c18239-static-analysis/build/checkerframework-output/main
    [javac] /home/jlewandowski/dev/cassandra/c18239-static-analysis/tools/fqltool/src/org/apache/cassandra/fqltool/ResultStore.java:102: error: [builder:required.method.not.called] @MustCall method removeAll may not have been invoked on field queues or any of its aliases.
    [javac]     private final @Owning ResourcesCollection<ChronicleQueue> queues;
    [javac]                                                               ^
    [javac]   The type of object is: org.apache.cassandra.utils.ResourcesCollection<net.openhft.chronicle.queue.ChronicleQueue>.
    [javac]   Reason for going out of scope:  @EnsuresCalledMethods written on MustCall methods doesn't contain method removeAll
    [javac] /home/jlewandowski/dev/cassandra/c18239-static-analysis/tools/fqltool/src/org/apache/cassandra/fqltool/ResultStore.java:103: error: [builder:required.method.not.called] @MustCall method removeAll may not have been invoked on field appenders or any of its aliases.
    [javac]     private final @Owning ResourcesCollection<ExcerptAppender> appenders;
    [javac]                                                                ^
    [javac]   The type of object is: org.apache.cassandra.utils.ResourcesCollection<net.openhft.chronicle.queue.ExcerptAppender>.
    [javac]   Reason for going out of scope:  @EnsuresCalledMethods written on MustCall methods doesn't contain method removeAll
    [javac] /home/jlewandowski/dev/cassandra/c18239-static-analysis/tools/fqltool/src/org/apache/cassandra/fqltool/ResultStore.java:104: error: [builder:required.method.not.called] @MustCall method close may not have been invoked on field queryStoreQueue or any of its aliases.
    [javac]     private final @Owning ChronicleQueue queryStoreQueue;
    [javac]                                          ^
    [javac]   The type of object is: net.openhft.chronicle.queue.ChronicleQueue.
    [javac]   Reason for going out of scope:  @EnsuresCalledMethods written on MustCall methods doesn't contain method close
    [javac] /home/jlewandowski/dev/cassandra/c18239-static-analysis/tools/fqltool/src/org/apache/cassandra/fqltool/ResultStore.java:105: error: [builder:required.method.not.called] @MustCall method close may not have been invoked on field queryStoreAppender or any of its aliases.
    [javac]     private final @Owning ExcerptAppender queryStoreAppender;
    [javac]                                           ^
    [javac]   The type of object is: net.openhft.chronicle.queue.ExcerptAppender.
    [javac]   Reason for going out of scope:  @EnsuresCalledMethods written on MustCall methods doesn't contain method close
    [javac] 4 errors
```

I don't quite understand what is going on and why this particular error is thrown. I would expect something more like `contracts.conditional.*`, but also not for all the fields. 


appendix:
```java
@InheritableMustCall({ "removeAll" })
public class ResourcesCollection<V> implements Iterable<V>
{
    private final Collection<V> collection;
    private final List<V> asList;

    public ResourcesCollection(Supplier<Collection<V>> collectionSupplier)
    {
        this.collection = collectionSupplier.get();
        this.asList = collection instanceof List ? (List<V>) collection : null;
    }

    @CreatesMustCallFor
    public boolean add(@Owning V value)
    {
        return collection.add(value);
    }

    @Override
    public ResourcesIterator<V> iterator()
    {
        return new ResourcesIterator<>(collection.iterator());
    }

    public <T> void removeAll(T initial, OwningBiFunction<T, V> consumer)
    {
        Iterator<V> iterator = collection.iterator();
        T acc = initial;
        while (iterator.hasNext())
        {
            V value = iterator.next();
            try
            {
                iterator.remove();
            }
            finally
            {
                acc = consumer.apply(acc, value);
            }
        }
    }

    public <T> void removeAll(OwningConsumer<V> consumer)
    {
        Iterator<V> iterator = collection.iterator();
        while (iterator.hasNext())
        {
            V value = iterator.next();
            try
            {
                iterator.remove();
            }
            finally
            {
                consumer.accept(value);
            }
        }
    }

    public @NotOwning V get(int index)
    {
        return asList != null ? asList.get(index) : Iterables.get(collection, index);
    }

    public boolean contains(V v)
    {
        return collection.contains(v);
    }

    public int size()
    {
        return collection.size();
    }

    public interface OwningConsumer<V> extends Consumer<V>
    {
        @Override
        void accept(@Owning V v);
    }

    public interface OwningBiFunction<T, V> extends BiFunction<T, V, T>
    {
        @Override
        T apply(T t, @Owning V v);
    }
}
```

--- a/checker-qual/src/main/java/org/checkerframework/checker/calledmethods/qual/EnsuresCalledMethods.java
+++ b/checker-qual/src/main/java/org/checkerframework/checker/calledmethods/qual/EnsuresCalledMethods.java
@@ -12,7 +12,10 @@ import org.checkerframework.framework.qual.QualifierArgument;
 
 /**
  * Indicates that the method, if it terminates successfully, always invokes the given methods on the
- * given expressions. This annotation is repeatable.
+ * given expressions. This annotation is repeatable, which means that users can write more than one
+ * instance of it on the same method (users should NOT manually write an
+ * {@code @EnsuresCalledMethods.List} annotation, which the checker will create from multiple copies
+ * of this annotation automatically).
  *
  * <p>Consider the following method:
  *
@@ -60,10 +63,10 @@ public @interface EnsuresCalledMethods {
   String[] methods();
 
   /**
-   * A wrapper annotation that makes the {@link EnsuresCalledMethods} annotation repeatable.
-   *
-   * <p>Programmers generally do not need to write this. It is created by Java when a programmer
-   * writes more than one {@link EnsuresCalledMethods} annotation at the same location.
+   * A wrapper annotation that makes the {@link EnsuresCalledMethods} annotation repeatable. This
+   * annotation is an implementation detail: programmers generally do not need to write this. It is
+   * created automatically by Java when a programmer writes more than one {@link
+   * EnsuresCalledMethods} annotation at the same location.
    */
   @Documented
   @Retention(RetentionPolicy.RUNTIME)

--- a/docs/manual/called-methods-checker.tex
+++ b/docs/manual/called-methods-checker.tex
@@ -193,7 +193,8 @@ There are also method annotations:
 \label{called-methods-ensurescalledmethods}
 \item[\refqualclass{checker/calledmethods/qual}{EnsuresCalledMethods}]
   This declaration annotation specifies a post-condition on a method, indicating the methods it
-  guarantees to be called.
+  guarantees to be called. This annotation is repeatable, meaning that you can write
+  multiple copies of it (with different arguments) on the same method, and the checker will check all of them.
 
   For example, this specification:
 

